### PR TITLE
NIFI-3055 StandardRecordWriter Can Throw UTFDataFormatException (1.x)

### PR DESCRIPTION
**Pull request by jskora**
* Updated StandardRecordWriter, even though it is now deprecated to consider the encoding behavior of java.io.DataOutputStream.writeUTF() and truncate string values such that the UTF representation will not be longer than that DataOutputStream's 64K UTF format limit.
* Updated the new SchemaRecordWriter class to similarly truncate long Strings that will be written as UTF.
* Add tests to confirm handling of large UTF strings and various edge conditions of UTF string handling.

Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by mosermw** [comment]
reviewing

File: nifi-commons/nifi-schema-utils/src/main/java/org/apache/nifi/repository/schema/SchemaRecordWriter.java [hunk 1]
```diff
@@ -136,4 +144,44 @@ private void writeFieldValue(final RecordField field, final Object value, final
...
     }
+
+    private void writeUTFLimited(final DataOutputStream out, final String utfString) throws IOException {
+        try {
+            out.writeUTF(utfString);
+        } catch (UTFDataFormatException e) {
+            final String truncated = utfString.substring(0, getCharsInUTFLength(utfString, MAX_ALLOWED_UTF_LENGTH));
+            logger.warn("Truncating UTF value!  Attempted to write string with char length {} and UTF length greater than "
+                            + "supported maximum allowed ({}), truncating to char length {}.",
+                    utfString.length(), MAX_ALLOWED_UTF_LENGTH, truncated.length());
```
Comment (mosermw) [inline nifi-commons/nifi-schema-utils/src/main/java/org/apache/nifi/repository/schema/SchemaRecordWriter.java]: Can we mention provenance in this message, such as "Truncating provenance record value"?  Does this message potentially mix char length and byte length, such as "Attempted to write string with char length 40000 and UTF length greater than supported maximum allowed (65535), truncating to char length 39000."?  Perhaps a simpler message such as "Attempted to store string with length 40000, truncating to 39000."

**Comment by jskora** [comment]
@mosermw I have updated the log message in both PRs.  Thanks!

**Comment by mosermw** [comment]
+1 looks good.  I like the unit tests!  I tested master branch before (exceptions galore and all partitions in flowfile repository was blacklisted) and after, we get appropriate WARN truncation messages in the logs.  Will squash and merge.
